From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HAOS-CN <haos-cn@local>
Date: Mon, 9 Jun 2026 00:00:00 +0800
Subject: [PATCH] rk3566-panther-x2: add rknpu support

Add NPU device nodes for Rockchip RK3566 Panther X2 board.
This includes:
- NPU QoS node for power domain state save/restore
- NPU power domain (PD_NPU) in the power controller
- NPU OPP table for devfreq dynamic frequency scaling
- RKNPU device node (non-IOMMU mode)
- RKNPU IOMMU node (disabled due to known v2 IOMMU limitations)

Based on: https://github.com/w568w/rknpu-module
---
 .../boot/dts/rockchip/rk3566-panther-x2.dts   | 81 +++++++++++++++++++
 1 file changed, 81 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk3566-panther-x2.dts b/arch/arm64/boot/dts/rockchip/rk3566-panther-x2.dts
index dc9f914a9..000000001 100644
--- a/arch/arm64/boot/dts/rockchip/rk3566-panther-x2.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3566-panther-x2.dts
@@ -156,6 +156,87 @@ wireless_wlan: wireless-wlan {
 
 };
 
+/* NPU QoS node for power domain state save/restore */
+qos_npu: qos@fde48000 {
+	compatible = "rockchip,rk3568-qos", "syscon";
+	reg = <0x0 0xfde48000 0x0 0x20>;
+};
+
+/* Add NPU power domain to the power controller */
+&power {
+	pd_npu: power-domain@RK3568_PD_NPU {
+		reg = <RK3568_PD_NPU>;
+		clocks = <&cru ACLK_NPU_PRE>,
+			 <&cru HCLK_NPU_PRE>,
+			 <&cru PCLK_NPU_PRE>;
+		pm_qos = <&qos_npu>;
+		#power-domain-cells = <0>;
+	};
+};
+
+/* NPU OPP table for devfreq */
+&{/} {
+	npu_opp_table: npu-opp-table {
+		compatible = "operating-points-v2";
+
+		opp-200000000 {
+			opp-hz = /bits/ 64 <200000000>;
+			opp-microvolt = <850000 850000 1000000>;
+		};
+		opp-297000000 {
+			opp-hz = /bits/ 64 <297000000>;
+			opp-microvolt = <850000 850000 1000000>;
+		};
+		opp-400000000 {
+			opp-hz = /bits/ 64 <400000000>;
+			opp-microvolt = <850000 850000 1000000>;
+		};
+		opp-600000000 {
+			opp-hz = /bits/ 64 <600000000>;
+			opp-microvolt = <850000 850000 1000000>;
+		};
+		opp-700000000 {
+			opp-hz = /bits/ 64 <700000000>;
+			opp-microvolt = <875000 875000 1000000>;
+		};
+		opp-800000000 {
+			opp-hz = /bits/ 64 <800000000>;
+			opp-microvolt = <925000 925000 1000000>;
+		};
+	};
+
+	rknpu: npu@fde40000 {
+		compatible = "rockchip,rk3568-rknpu", "rockchip,rknpu";
+		reg = <0x00 0xfde40000 0x00 0x10000>;
+		interrupts = <GIC_SPI 151 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "npu_irq";
+		clocks = <&scmi_clk SCMI_CLK_NPU>,
+			 <&cru CLK_NPU>,
+			 <&cru ACLK_NPU>,
+			 <&cru HCLK_NPU>;
+		clock-names = "scmi_clk", "clk", "aclk", "hclk";
+		assigned-clocks = <&cru CLK_NPU>;
+		assigned-clock-rates = <600000000>;
+		resets = <&cru SRST_A_NPU>, <&cru SRST_H_NPU>;
+		reset-names = "srst_a", "srst_h";
+		power-domains = <&power RK3568_PD_NPU>;
+		operating-points-v2 = <&npu_opp_table>;
+		rknpu-supply = <&vdd_npu>;
+		status = "okay";
+	};
+
+	/* IOMMU for NPU — disabled (IOMMU v2 has 4GB limit issue) */
+	rknpu_mmu: iommu@fde4b000 {
+		compatible = "rockchip,rk3568-iommu";
+		reg = <0x00 0xfde4b000 0x00 0x40>;
+		interrupts = <GIC_SPI 150 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "rknpu_mmu";
+		clocks = <&cru ACLK_NPU>, <&cru HCLK_NPU>;
+		clock-names = "aclk", "iface";
+		power-domains = <&power RK3568_PD_NPU>;
+		#iommu-cells = <0>;
+		status = "disabled";
+	};
+};
+
 &cpu0 {
 	cpu-supply = <&vdd_cpu>;
 };
-- 
2.45.2

